adapters/dkb-adapter/src/main/java/de/adorsys/xs2a/gateway/service/impl/DkbAccessTokenService.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2018-2018 adorsys GmbH & Co KG
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package de.adorsys.xs2a.gateway.service.impl;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import de.adorsys.xs2a.gateway.http.HttpClient;
+import de.adorsys.xs2a.gateway.http.JsonMapper;
+import de.adorsys.xs2a.gateway.security.AccessTokenException;
+import de.adorsys.xs2a.gateway.security.AccessTokenService;
+import de.adorsys.xs2a.gateway.service.GeneralResponse;
+
+import java.time.LocalDateTime;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+
+public class DkbAccessTokenService implements AccessTokenService {
+
+    public static final String DKB_TOKEN_CONSUMER_KEY_PROPERTY = "dkb.token.consumer_key";
+    public static final String DKB_TOKEN_CONSUMER_SECRET_PROPERTY = "dkb.token.consumer_secret";
+    private static final String DKB_TOKEN_URL_PROPERTY = "dkb.token.url";
+    private static final String DKB_TOKEN_SECONDS_BEFORE_TOKEN_EXPIRATION_PROPERTY = "dkb.token.seconds_before_token_expiration";
+
+    private static final String DEFAULT_SECONDS_BEFORE_TOKEN_EXPIRATION = "60";
+    private static final String DEFAULT_TOKEN_URL = "https://api.dkb.de/token";
+
+    private static AccessTokenService instance;
+    private static Map<String, String> headers;
+    private static String tokenUrl;
+    private static int secondsBeforeTokenExpiration;
+
+    private HttpClient httpClient;
+
+    private AccessToken accessToken;
+
+    private JsonMapper jsonMapper;
+
+
+    private DkbAccessTokenService() {
+        jsonMapper = new JsonMapper();
+        setHttpClient(HttpClient.newHttpClient());
+    }
+
+    public static AccessTokenService getInstance() {
+        if (instance == null) {
+            instance = new DkbAccessTokenService();
```

Comment:
What do you think: shouldn't we be afraid of race conditions here? I mean, it is possible, that more than 1 instances could be created. I think, maybe lazy initialization is not needed here. What is your opinion?

Reply:
Thanks, it makes sense